File: Sources/DotNet/Woopsa/Services/SubscriptionService/SubscriptionService.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Woopsa
{
    public class SubscriptionService : WoopsaObject, IWoopsaServiceSubscription
    {
        public readonly TimeSpan ChannelTimedOutCheckInterval = TimeSpan.FromSeconds(60);

        #region Constructors

        public SubscriptionService(WoopsaContainer container)
            : base(container, WoopsaServiceSubscriptionConst.WoopsaServiceSubscriptionName)
        {
            _createSubscriptionChannel = new WoopsaMethod(
                this,
                WoopsaServiceSubscriptionConst.WoopsaCreateSubscriptionChannel,
                WoopsaValueType.Integer,
                new WoopsaMethodArgumentInfo[] { new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaNotificationQueueSize, WoopsaValueType.Integer) },
                args => CreateSubscriptionChannel(args.First())
            );

            _registerSubscription = new WoopsaMethod(
                this,
                WoopsaServiceSubscriptionConst.WoopsaRegisterSubscription,
                WoopsaValueType.Integer,
                new WoopsaMethodArgumentInfo[] {
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaSubscriptionChannel, WoopsaValueType.Integer),
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaPropertyLink, WoopsaValueType.WoopsaLink),
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaMonitorInterval, WoopsaValueType.TimeSpan),
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaPublishInterval, WoopsaValueType.TimeSpan)
                },
                args =>
                {
                    var arguments = args as IWoopsaValue[] ?? args.ToArray();
                    return RegisterSubscription(arguments[0], arguments[1], arguments[2], arguments[3]);
                });

            _unregisterSubscription = new WoopsaMethod(
                this,
                WoopsaServiceSubscriptionConst.WoopsaUnregisterSubscription,
                WoopsaValueType.Logical,
                new WoopsaMethodArgumentInfo[] {
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaSubscriptionChannel, WoopsaValueType.Integer),
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaSubscriptionId, WoopsaValueType.Integer)
                },
                args =>
                {
                    var arguments = args as IWoopsaValue[] ?? args.ToArray();
                    return UnregisterSubscription(arguments[0], arguments[1]);
                });

            _waitNotification = new WoopsaMethod(
                this,
                WoopsaServiceSubscriptionConst.WoopsaWaitNotification,
                WoopsaValueType.JsonData,
                new WoopsaMethodArgumentInfo[] {
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaSubscriptionChannel, WoopsaValueType.Integer),
                    new WoopsaMethodArgumentInfo(WoopsaServiceSubscriptionConst.WoopsaLastNotificationId, WoopsaValueType.Integer)
                },
                args =>
                {
                    var arguments = args as IWoopsaValue[] ?? args.ToArray();
                    return WaitNotification(arguments[0], arguments[1]);
                });

            _timerCheckChannelTimedOut = new LightWeightTimer(ChannelTimedOutCheckInterval);
            _timerCheckChannelTimedOut.Elapsed += _timerCheckChannelTimedOut_Elapsed;
            _timerCheckChannelTimedOut.IsEnabled = true;
        }

        private void _timerCheckChannelTimedOut_Elapsed(object sender, EventArgs e)
        {
            WoopsaSubscriptionChannel[] timedoutChannels;
            lock(_channels)
            {
                timedoutChannels = _channels.Values.Where((item) => item.ClientTimedOut).ToArray();
            }
            foreach (var item in timedoutChannels)
            {
                lock(_channels)
                    _channels.Remove(item.Id);
                item.Dispose();
            }
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);
            if (disposing)
            {
                if (_timerCheckChannelTimedOut != null)
                {
                    _timerCheckChannelTimedOut.Dispose();
                    _timerCheckChannelTimedOut = null;
                }
                if (_channels != null)
                {
                    foreach (var item in _channels.Values)
                        item.Dispose();
                    _channels = null;
                }
            }
        }

        #endregion

        #region IWoopsaServiceSubscription

        public WoopsaValue CreateSubscriptionChannel(IWoopsaValue notificationQueueSize)
        {
            var channel = new WoopsaSubscriptionChannel(notificationQueueSize.ToInt32());
            lock (_channels)
                _channels.Add(channel.Id, channel);
            return new WoopsaValue(channel.Id);
        }

        public WoopsaValue RegisterSubscription(IWoopsaValue subscriptionChannel, IWoopsaValue woopsaPropertyLink, IWoopsaValue monitorInterval, IWoopsaValue publishInterval)
        {
            int channelId = subscriptionChannel.ToInt32();
            lock (_channels)
                if (_channels.ContainsKey(channelId))
                    return _channels[channelId].RegisterSubscription(Owner, woopsaPropertyLink, monitorInterval.ToTimeSpan(), publishInterval.ToTimeSpan());
                else
                    throw new WoopsaException(string.Format("Tried to register a subscription on channel with id={0} that does not exist", subscriptionChannel.ToInt32()));
        }

        // Logical, true if the subscription has been found and successfully unregistered
        public WoopsaValue UnregisterSubscription(IWoopsaValue subscriptionChannel, IWoopsaValue subscriptionId)
        {
            int channelId = subscriptionChannel.ToInt32();
            lock (_channels)
                if (_channels.ContainsKey(channelId))
                    return new WoopsaValue(_channels[channelId].UnregisterSubscription(subscriptionId.ToInt32()));
                else
                    throw new WoopsaException(string.Format("Tried to unregister a subscription on channel with id={0} that does not exist", subscriptionChannel.ToInt32()));
        }

        public WoopsaValue WaitNotification(IWoopsaValue subscriptionChannel, IWoopsaValue lastNotificationId)
        {
            int channelId = subscriptionChannel.ToInt32();
            int notificationId = lastNotificationId.ToInt32();
            WoopsaSubscriptionChannel channel;
            lock (_channels)
                if (_channels.ContainsKey(channelId))
                    channel = _channels[channelId];
                else
                    channel = null;
            if (channel != null)
            {
                IWoopsaNotifications notifications = channel.WaitNotification(WoopsaServiceSubscriptionConst.NotificationTimeoutInterval, notificationId);
                return WoopsaValue.WoopsaJsonData(notifications.Serialize());
            }
            else
                throw new WoopsaInvalidSubscriptionChannelException(string.Format("Tried to call WaitNotification on channel with id={0} that does not exist", channelId));
        }

        #endregion

        #region Private Members

        private WoopsaMethod _createSubscriptionChannel;
        private WoopsaMethod _registerSubscription;
        private WoopsaMethod _unregisterSubscription;
        private WoopsaMethod _waitNotification;
        private Dictionary<int, WoopsaSubscriptionChannel> _channels = new Dictionary<int, WoopsaSubscriptionChannel>();
        private LightWeightTimer _timerCheckChannelTimedOut;

        #endregion
    }
}

File: Sources/JavaScript/nodejs/server/subscription-service.js
var types = require('./types');
var exceptions = require('./exceptions');
var subscriptionChannel = require('./subscription-channel');

var SubscriptionService = function SubscriptionService(){
    this._createSubscriptionChannel = new types.WoopsaMethod("CreateSubscriptionChannel", "Integer", createSubscriptionChannel, [
        {"NotificationQueueSize": "Integer"}
    ]);

    this._registerSubscription = new types.WoopsaMethod("RegisterSubscription", "Integer", registerSubscription, [
        {"SubscriptionChannel": "Integer"},
        {"PropertyLink": "WoopsaLink"},
        {"MonitorInterval": "TimeSpan"},
        {"PublishInterval": "TimeSpan"}
    ]);

    this._unregisterSubscription = new types.WoopsaMethod("UnregisterSubscription", "Logical", unregisterSubscription, [
        {"SubscriptionChannel": "Integer"},
        {"SubscriptionId": "Integer"}
    ]);

    this._waitNotification = new types.WoopsaMethod("WaitNotification", "JsonData", waitNotification, [
        {"SubscriptionChannel": "Integer"},
        //{"LastNotificationId": "Integer"}
    ]);
    // Makes the server pass us the Response object
    this._waitNotification.invokeAsync = function (arguments, done){
        waitNotification.apply(this, arguments.concat([done]));
    }

    var subscriptionChannels = {};
    var lastSubscriptionChannelId = 0;

    function createSubscriptionChannel(notificationQueueSize){
        var newChannel = new subscriptionChannel.SubscriptionChannel(notificationQueueSize);
        subscriptionChannels[++lastSubscriptionChannelId] = newChannel;
        return lastSubscriptionChannelId;
    }

    function registerSubscription(subscriptionChannel, propertyLink, monitorInterval, publishInterval){
        if ( typeof subscriptionChannels[subscriptionChannel] === 'undefined' ){
            throw new exceptions.WoopsaInvalidSubscriptionChannelException("Tried to call RegisterSubscription on channel with id=" + subscriptionChannel + " that does not exist.");
        }
        var channel = subscriptionChannels[subscriptionChannel];
        return channel.registerSubscription(propertyLink, monitorInterval, publishInterval);
    }

    function unregisterSubscription(subscriptionChannel, subscriptionId){
        if ( typeof subscriptionChannels[subscriptionChannel] === 'undefined' ){
            throw new exceptions.WoopsaInvalidSubscriptionChannelException("Tried to call UnregisterSubscription on channel with id=" + subscriptionChannel + " that does not exist.");
        }
        var channel = subscriptionChannels[subscriptionChannel];
        return channel.unregisterSubscription(subscriptionId);
    }

    function waitNotification(subscriptionChannel, done){
        // if ( typeof subscriptionChannels[subscriptionChannel] === 'undefined' ){
        //    throw new exceptions.WoopsaInvalidSubscriptionChannelException("Tried to call UnregisterSubscription on channel with id=" + subscriptionChannel + " that does not exist.");
        // }
        // var channel = subscriptionChannels[subscriptionChannel];
        setTimeout(function (){
            done({Something:"lol"});
        }, 2000);
        // return channel.waitNotification(1, done);
    }
}

SubscriptionService.prototype = {
    getName: function (){
        return "SubscriptionService";
    },
    getProperties: function (){
        return [];
    },
    getMethods: function (){
        return [
            this._createSubscriptionChannel,
            this._registerSubscription,
            this._unregisterSubscription,
            this._waitNotification,
        ];
    },
    getItems: function (){
        return [];
    }
}

exports.SubscriptionService = SubscriptionService;